server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsRemoteDeliveryPermanentError.java

```diff
@@ -0,0 +1,53 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.transport.matchers;
+
+import static org.apache.james.transport.mailets.remote.delivery.Bouncer.IS_DELIVERY_PERMANENT_ERROR;
+
+import java.util.Collection;
+
+import org.apache.james.core.MailAddress;
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.GenericMatcher;
+
+import com.google.common.collect.ImmutableList;
+
+/**
+ * <p>
+ * Checks if the mail has a permanent remote delivery failure attribute
+ * </p>
+ *
+ * Example:
+ *
+ * <pre><code>
+ * &lt;mailet match=&quot;IsRemoteDeliveryPermanentError&quot; class=&quot;&lt;any-class&gt;&quot;/&gt;
+ * </code></pre>
+ */
+public class IsRemoteDeliveryPermanentError extends GenericMatcher {
+
+    public static final AttributeName IS_DELIVERY_PERMANENT_ERROR = AttributeName.of("is-delivery-permanent-error");
```

Comment:
Please reuse the error code attribute added in https://github.com/linagora/james-project/pull/3552 (delivery-error-code), don't add yet-another attribute!

Reply:
You need to check in the mailet if the code is a permanent error or a temporary one right? What codes are considered permanent then?

Reply:
or we will define another attribute for temporary errors? I don't really get this...

Reply:
To be explicit 5xx -> permanent, 4xx -> we can retry

Souce https://www.wikiwand.com/en/List_of_SMTP_server_return_codes#/4xx_Persistent_transient_failure

Reply:
Permanent error could include other kinds of error besides SMTP, so adding another attribute would be more convenient.